README.md
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](LICENSE)
[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-v1.4%20adopted-ff69b4.svg)](CODE_OF_CONDUCT.md)


# JeLLyFysh

The JeLLyFysh Python application implements the event-chain Monte Carlo algorithm (ECMC), an event-driven irreversible 
Markov-chain Monte Carlo algorithm for classical *N*-body simulations in statistical mechanics, biophysics and 
electrochemistry. The application's architecture closely mirrors the mathematical formulation of ECMC. For a detailed 
description of the design of this application, see [\[Hoellmer2020\]](https://doi.org/10.1016/j.cpc.2020.107168). For a 
closely connected discussion of ECMC, see [\[Faulkner2018\]](https://doi.org/10.1063/1.5036638). (For more information 
on the references, see the [References.bib](References.bib) file.)

## Installing

The JeLLyFysh Python application can be executed with any Python implementation which supports Python version >= 3.6. It 
is tested with cPython and [PyPy](http://pypy.org). We recommend using the Python 3.8 compatible PyPy3.8 v7.3.7 
(see https://www.pypy.org/download.html), which yields a significant higher performance of the JeLLyFysh application.

The JeLLyFysh application is installed with the help of setuptools (see [setup.py](setup.py)). After you have downloaded 
and uncompressed the JeLLyFysh source, or cloned it from git, first choose a Python interpreter that supports version 
3.6 or greater. We recommend installing the JeLLyFysh application into a virtual environment (e.g., using virtualenv, 
venv, pyenv, ...). This documentation uses the command `pypy3` to call the Python interpreter of the virtual 
environment. If you use a different Python interpreter, change the code accordingly.

1.  Ensure that pip is installed by running

    ```console
    pypy3 -m pip --help
    ```
   
    If this command does not show pip's help message, install it (see, e.g., 
    https://pip.pypa.io/en/stable/installing/). Moreover, the C extensions of the JeLLyFysh application require a C 
    compiler (for example, gcc or clang).


2.  Ensure that pip, setuptools, and wheel are installed and on the latest version by running

    ```console
    pypy3 -m pip install --upgrade pip setuptools wheel
    ```

   
3. You are now ready to install the JeLLyFysh application. To do so, navigate into the directory of this file (and the 
    relevant [setup.py](setup.py) script), and execute

    ```console
    pypy3 -m pip install .
    ```
   
    This will first install the required dependencies of the JeLLyFysh application (see
    [EXTERNAL_DEPENDENCIES.md](EXTERNAL_DEPENDENCIES.md)). Afterwards, it compiles the C extensions of the JeLLyFysh 
    application, and exports the [`jellyfysh`](jellyfysh) package to the site-packages directory of your interpreter. 
    Finally, it creates the executables described in the [section "Using JeLLyFysh"](#using-jellyfysh).


4.  A limited amount of modules depends on additional Python packages that are not installed by setuptools. Take a look 
    at the file [EXTERNAL_DEPENDENCIES.md](EXTERNAL_DEPENDENCIES.md) to see which parts of the JeLLyFysh application can 
    only be used after additional packages have been installed manually with the help of pip.

### Uninstalling

In order to uninstall the JeLLyFysh application, use

```console
pypy3 -m pip uninstall jellyfysh
```

### Editable mode

Using the command `pypy3 -m pip install .` to install the JeLLyFysh application basically copies and compiles the source 
code into the site-packages directory of your Python interpreter. Therefore, you would need to reinstall the JeLLyFysh 
application anytime source code was changed. Fortunately, setuptools provides an editable mode. This installs the 
JeLLyFysh application without copying any files. Any changes in source code are directly active without reinstallation. 
To install the JeLLyFysh application in editable mode, use

```console
pypy3 -m pip install -e .
```

Of course, a C extension must still be recompiled after any changes to the corresponding C code. This can be achieved by
using

```console
    pypy3 setup.py build_ext -i
```

Uninstalling the JeLLyFysh application in editable mode works as described above.

## Running the tests

To run the tests, simply enter the [`unittests`](unittests) directory and run the 
[`run_tests.py`](unittests/run_tests.py) script with your Python interpreter. These tests may take a while.

## Using JeLLyFysh

The installation process described in the [section "Installing"](#installing) creates three executables that are 
described in the following. Note that you might have to add the directory where pip installs executables to your `PATH` 
environment variable in order to access these executables.

### 1. jellyfysh 

The `jellyfysh` executable runs the JeLLyFysh application and relies on the [`run.py`](jellyfysh/run.py) script, which 
is located in the [`jellyfysh`](jellyfysh) package. (The `jellyfysh` executable is therefore analogous to running the 
[`run.py`](jellyfysh/run.py) script with your Python interpreter.)

The user interface for each run of the JeLLyFysh application consists in a configuration file that is an argument of the 
`jellyfysh` executable. Configuration files should follow the [INI-file format](https://en.wikipedia.org/wiki/INI_file).

The `jellyfysh` executable expects the path to the configuration file as the first positional argument. The executable 
also takes optional arguments. These are:

- `-h`, `--help`: Show the help message and exit.
- `-V`, `--version`: Show program's version number and exit.
- `-v`, `--verbose`: Increase verbosity of logging messages (multiple -v options increase the verbosity, the maximum is 
2).
- `-l LOGFILE`, `--logfile LOGFILE`: Specify the logging file. 

A configuration file is composed of sections that each correspond to a class of the JeLLyFysh application. The only
required section for the run script is

```INI
[Run]
mediator = some_mediator
setting = some_setting
```

`some_mediator` corresponds to the used mediator in the run. The mediator serves as a central hub in the application 
and also hosts the iteration loop over the legs of the continuous-time evolution of ECMC. The two possible mediators are
[`single_process_mediator`](jellyfysh/mediator/single_process_mediator.py), which runs the application in a single process, 
and [`multi_process_mediator`](jellyfysh/mediator/multi_process_mediator/multi_process_mediator.py), which calculates the 
events of ECMC in separate processes. Both are located in the [`jellyfysh.mediator`](jellyfysh/mediator) package. 
Currently, we recommend the single-process version.

`setting` specifies the *NVT* physical parameters of the run. The two possible settings are 
[`hypercubic_setting`](jellyfysh/setting/hypercubic_setting.py) and 
[`hypercuboid_setting`](jellyfysh/setting/hypercuboid_setting.py) (both located in the 
[`jellyfysh.setting`](jellyfysh/setting) package).

The following sections of the configuration file choose the parameters in the `__init__` methods of the mediator and the 
setting. Each section contains pairs of properties and values. The property corresponds to the name of the argument in 
the `__init__` method of the given class, and its value provides the arguments. Properties and values should be given in 
snake_case, sections in CamelCase. For a detailed description of the JeLLyFysh factory, which parses the configuration 
file and constructs the specified classes, see the [FACTORY.md](FACTORY.md) file. There, also a detailed example for the 
interplay between the factory and a configuration file is given. All classes of JeLLyFysh are documented with docstrings 
to clarify their usage. For a general overview of the parts of JeLLyFysh, see 
[\[Hoellmer2020\]](https://doi.org/10.1016/j.cpc.2020.107168).

Instances of the event-handler classes in the JeLLyFysh application compute candidate events. Hints on how to implement 
your own event handlers are given in the [HOWTO_EVENT_HANDLER.md](HOWTO_EVENT_HANDLER.md) file (see 
[\[Hoellmer2020\]](https://doi.org/10.1016/j.cpc.2020.107168) for details).

### 2. jellyfysh-examples

As a starting point into the JeLLyFysh application, we provide exemplary configuration files. The `jellyfysh-examples` 
executable (which relies on the [`create_examples.py`](jellyfysh/create_examples.py) script in the 
[`jellyfysh`](jellyfysh) package) creates a `jellyfysh-examples` directory in the current working directory. This 
directory contains several exemplary configuration files. The configuration files in the 
[`jellyfysh-examples/config_files/2018_JCP_149_064113`](jellyfysh/config_files/2018_JCP_149_064113) directory are 
described in detail in the "Cookbook" section (Section 5) of 
[\[Hoellmer2020\]](https://doi.org/10.1016/j.cpc.2020.107168). They generate output in the 
[`jellyfysh-examples/output/2018_JCP_149_064113`](jellyfysh/output/2018_JCP_149_064113) directory. This output directory 
also contains plotting scripts that compare the results of the JeLLyFysh application with reference data. (The plotting 
scripts, however, rely on additional external dependencies that have to be installed manually, see 
[EXTERNAL_DEPENDENCIES.md](EXTERNAL_DEPENDENCIES.md).) More examples can be found 
in the [`jellyfysh-examples/config_files/hard_disk_dipoles`](jellyfysh/config_files/hard_disk_dipoles) directory that 
generate output in the [`jellyfysh-examples/output/hard_disk_dipoles`](jellyfysh/output/hard_disk_dipoles) directory.

For your first run of the JeLLyFysh application, you may use the following commands:

```console
jellyfysh-examples
cd jellyfysh-examples
jellyfysh config_files/2018_JCP_149_064113/coulomb_atoms/power_bounded.ini
```

This simulates two charged point masses in a three-dimensional cubic simulation box with periodic boundary conditions. 
The generated file `output/2018_JCP_149_064113/coulomb_atoms/SamplesOfSeparation_PowerBounded.dat` contains samples of 
the separation between the point masses. Only if you installed the necessary external dependencies described in 
[EXTERNAL_DEPENDENCIES.md](EXTERNAL_DEPENDENCIES.md), you can plot the data and compare it to reference data from a 
reversible Monte Carlo simulation by running the following commands (again assuming that `pypy3` calls your Python 
interpreter):

```console
cd output/2018_JCP_149_064113/coulomb_atoms
pypy3 plot_histogram_coulomb_atoms.py 
```

This creates a plot and stores it in the `coulomb_atoms.pdf` file.

### 3. jellyfysh-resume

The `jellyfysh-resume` executable (which relies on the [`resume.py`](jellyfysh/resume.py) script in the 
[`jellyfysh`](jellyfysh) package) resumes a dumped run of the JeLLyFysh application. Such a dumped run is generated by 
including the [`DumpingOutputHandler`](jellyfysh/input_output_handler/output_handler/dumping_output_handler.py) in 
a run of the JeLLyFysh application.

We provide an exemplary configuration file that dumps a run of the JeLLyFysh application. To run this example, use the 
following commands:

```console
jellyfysh-examples
cd jellyfysh-examples
jellyfysh config_files/2018_JCP_149_064113/coulomb_atoms/power_bounded_dump.ini
```

The `DumpingOutputHandler` pickles all objects of the running simulation and dumps them into a file whose name depends 
on the used Python interpreter as `dump_PythonImplementation_PythonVersion.dat`. You can resume the run from the moment 
where the dump was created by using

```console
jellyfysh-resume dump_PythonImplementation_PythonVersion.dat
```

The `jellyfysh-resume` executable is mainly used for debugging.

## Contributing

As an open-source project, the JeLLyFysh organization solicits contributions from the community. Please read 
the [contribution guideline](CONTRIBUTING.md) for details.

If you find a bug, please raise an Issue here on GitHub to let us know.

Please note that this project is released with the Contributor Covenant [code of conduct](CODE_OF_CONDUCT.md). By 
participating in this project you agree to abide by its terms. Report unacceptable behavior to 
[werner.krauth@ens.fr](mailto:werner.krauth@ens.fr).

## Versioning

Versioning of the JeLLyFysh project adopts two-to-four-field version numbers defined as Milestone.Feature.AddOn.Patch. 
The current version 1.0 represents the first development milestone which reproduces published data in 
[\[Faulkner2018\]](https://doi.org/10.1063/1.5036638). Patches and bugfixes of this version will be given number 
1.0.0.1, 1.0.0.2 etc. New configuration files and required extensions are expected to lead to versions 1.0.1, 1.0.2 etc. 
In JeLLyFysh development, two-field versions (2.0, 3.0, etc.) may introduce incompatible code, while three- and 
four-field version numbers are intended to be backward compatible.

## Authors

Check the [AUTHORS.md](AUTHORS.md) file to see who participated in this project.

## License

This project is licensed under the GNU General Public License, version 3 (see the [LICENSE](LICENSE) file).

## Contact

If you have questions regarding the JeLLyFysh application, just raise an issue here on GitHub. We are happy to help you!

## Citation

If you use JeLLyFysh in published work, please cite the following reference (see [Hoellmer2020] in 
[References.bib](References.bib)):

Philipp Höllmer, Liang Qin, Michael F. Faulkner, A. C. Maggs, and Werner Krauth,  
JeLLyFysh-Version1.0 -- a Python application for all-atom event-chain Monte Carlo,  
Computer Physics Communications, Volume 253, 107168 (2020), https://doi.org/10.1016/j.cpc.2020.107168.


CHANGELOG.md
# Changelog

All notable changes to the JeLLyFysh application will be documented in this file.  The latest version comes first, and
the release dates are written in the [ISO 8601 format](https://en.wikipedia.org/wiki/ISO_8601) [YYYY]-[MM]-[DD].

The version numbers of the JeLLyFysh application adopt [semantic versioning](http://semver.org) with two-to-four-field 
version numbers defined as Milestone.Feature.AddOn.Patch. We distinguish between backwards compatible changes, and 
backwards incompatible changes. Here, compatible refers to changes that do not break existing configuration files. 
In JeLLyFysh development, two-field versions (2.0, 3.0, etc.) may introduce incompatible code, while three- and 
four-field version numbers are intended to be backward compatible (or to yield only minor changes in existing 
configuration files). Any version in this changelog explicitly reports which changes in existing configuration files are 
required.

## Version1.1 &ndash; 2022-02-07

### Major Changes

- The arXiv reference [\[Hoellmer2019\]](https://arxiv.org/abs/1907.12502) was replaced by the journal reference 
  [\[Hoellmer2020\]](https://doi.org/10.1016/j.cpc.2020.107168). If you use the JeLLyFysh application in published work,
  please cite the updated reference (see [References.bib](References.bib)).

- New installation process via setuptools (see [`setup.py`](setup.py)). All source code is now
  part of a `jellyfysh` Python package. The new installation process is described in the [README.md](README.md) file. 
  This has the following advantages:
  - Automatically installs all required external dependencies (see 
    [EXTERNAL_DEPENDENCIES.md](EXTERNAL_DEPENDENCIES.md)).
  - Simplifies the usage and installation of C extensions.
  - Enables the creation of executables like `jellyfysh` that replace explicitly running Python scripts like 
    [`run.py`](jellyfysh/run.py) with a Python interpreter.
  - Allows for simple access to the `jellyfysh` source code without manipulating `PYTHONPATH` after installation.

- Installation of three executables `jellyfysh`, `jellyfysh-resume`, and `jellyfysh-examples` via setuptools. They are 
  described in detail in the [README.md](README.md) file and simplify using the JeLLyFysh application.

- Some parts of JeLLyFysh are now implemented in C using [CFFI](https://cffi.readthedocs.io/en/latest/) 
  for the interface (which is the [recommended way for PyPy](https://doc.pypy.org/en/latest/extending.html)). This 
  increases the performance of JeLLyFysh.

- The JeLLyFysh application now only supports Python implementations that support Python versions >= 3.6.

- Added support for ECMC simulations of hard spheres, and of hard dipoles consisting of two tethered hard spheres (which 
  is inspired by a recent work of some members of the JeLLyFysh organization, see 
  [\[Hoellmer2021\]](https://arxiv.org/abs/2111.11943) in [References.bib](References.bib)). JeLLyFysh includes new 
  [exemplary configuration files](jellyfysh/config_files/hard_disk_dipoles) for ECMC simulations of two-dimensional 
  tethered hard-disk dipoles together with reference data and plotting scripts.

- Support for general velocities of active units (that are not only aligned with the coordinate axes as in the previous 
  version of JeLLyFysh). This is possible of the following major changes in the [`potential`](jellyfysh/potential) 
  package:
  - The `derivative` method of the abstract [`Potential`](jellyfysh/potential/potential.py) class now receives the full 
    velocity of the active unit as the first argument, and returns the directional time derivative. 
  - The `displacement` method of the abstract [`InvertiblePotential`](jellyfysh/potential/potential.py) now receives the 
    full velocity of the active unit as the first argument, and returns the time displacement.

- Added support for sequential ECMC which rotates the velocity vector of the independent active unit by a constant 
  angle increment in each end-of-chain event (see [\[Hoellmer2021\]](https://arxiv.org/abs/2111.11943)). Some new
  [exemplary configuration files](jellyfysh/config_files/hard_disk_dipoles) for ECMC simulations of two-dimensional 
  tethered hard-disk dipoles use this sequential ECMC.

- Candidate event times are now stored and compared by using a new [`Time`](jellyfysh/base/time.py) class that
  represents a time as the quotient and remainder of an integer division with one.

  This change is necessary because the candidate event times and time stamps of active units always increase in 
  simulations of JeLLyFysh. At the same time, event 
  handlers usually first compute a time displacement and then add this time displacement to a time stamp of an active 
  unit to determine the candidate event time. Here, the time displacements stay in the same order of magnitude during a 
  simulation. If floats are used to store the involved time stamp and resulting candidate event time, events with very 
  close time displacements may yield an equal candidate event time if the time stamp is already rather large. This leads 
  to a decreasing precision during the simulation. This problem is solved by the `Time` class because the precision of 
  the remainder stays constant during a simulation.

###  Required Changes of Configuration Files

- A recent work by some members of the JeLLyFysh organization (see 
  [\[Hoellmer2021\]](https://arxiv.org/abs/2111.11943) in [References.bib](References.bib)) showed that the independent
  active unit should be sampled randomly after an end-of-chain event. Keeping the same unit independent active before 
  and after an end-of-chain event can break irreducibility in simple models. Therefore, the
  `SameActivePeriodicDirectionEndOfChainEventHandler` was removed and should be replaced by the
  [`SingleIndependentActivePeriodicDirectionEndOfChainEventHandler`](jellyfysh/event_handler/single_independent_active_periodic_direction_end_of_chain_event_handler.py).

jellyfysh/run.py
# JeLLFysh - a Python application for all-atom event-chain Monte Carlo - https://github.com/jellyfysh
# Copyright (C) 2019, 2022 The JeLLyFysh organization
# (See the AUTHORS.md file for the full list of authors.)
#
# This file is part of JeLLyFysh.
#
# JeLLyFysh is free software: you can redistribute it and/or modify it under the terms of the GNU General Public
# License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later
# version.
#
# JeLLyFysh is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
# warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with JeLLyFysh in the LICENSE file.
# If not, see <https://www.gnu.org/licenses/>.
#
# If you use JeLLyFysh in published work, please cite the following reference (see [Hoellmer2020] in References.bib):
# Philipp Hoellmer, Liang Qin, Michael F. Faulkner, A. C. Maggs, and Werner Krauth,
# JeLLyFysh-Version1.0 -- a Python application for all-atom event-chain Monte Carlo,
# Computer Physics Communications, Volume 253, 107168 (2020), https://doi.org/10.1016/j.cpc.2020.107168.
#
"""Executable script which runs the JeLLyFysh application based on a configuration file."""
from argparse import ArgumentParser, Namespace
from configparser import ConfigParser
import logging
import platform
import sys
import time
from typing import Sequence
from jellyfysh.base.exceptions import EndOfRun
from jellyfysh.base import factory
from jellyfysh.base.strings import to_camel_case
from jellyfysh.base.uuid import get_uuid
import jellyfysh.version as version


def add_general_parser_arguments(parser: ArgumentParser) -> None:
    """
    Add parser arguments to the command line argument parser.

    This method adds the following possible arguments:
    1. --version, -V: Print the version of the application and exit.
    2. --verbose, -v: Increase verbosity of logging messages. Multiple -v options increase the verbosity. The maximum
    is 2.
    3. --logfile LOGFILE, -l LOGFILE: Specify the logging file.
    Per default, also the following argument is added:
    4. --help, -h: Show the help message and exit.

    Parameters
    ----------
    parser : argparse.ArgumentParser
        The argument parser.
    """
    parser.add_argument("-V", "--version", action="version", version="JeLLyFysh application version "
                                                                     + version.__version__)
    parser.add_argument("-v", "--verbose", action="count",
                        help="increase verbosity of logging messages "
                             "(multiple -v options increase the verbosity, the maximum is 2)")
    parser.add_argument("-l", "--logfile", action="store", help="specify the logging file")


def parse_options(args: Sequence[str]) -> Namespace:
    """
    Convert argument strings to objects and assign them as attributes of the argparse namespace. Return the populated
    namespace.

    The argument strings can for example be sys.argv[1:] in order to parse the command line arguments. The argument
    parser parses the arguments specified in the add_general_parser_arguments function. This function also adds the
    configuration file as a required positional argument.

    Parameters
    ----------
    args : Sequence[str]
        The argument strings.

    Returns
    -------
    argparse.Namespace
        The populated argparse namespace.
    """
    parser = ArgumentParser(description="Run the JeLLyFysh application based on a configuration file.")
    parser.add_argument("config_file", help="specify the path to the configuration file of the run")
    add_general_parser_arguments(parser)
    return parser.parse_args(args)


def set_up_logging(parsed_arguments: Namespace) -> logging.Logger:
    """
    Set up the logging based on the populated argparse namespace.

    The level of the root logger is set based on the number of -v arguments parsed. For zero -v arguments, it is set
    to logging.WARNING, for one to logging.INFO and for two to logging.DEBUG. If a log file was specified in the
    parsed arguments, the logging information is written into that file.

    Parameters
    ----------
    parsed_arguments : argparse.Namespace
        The populated argparse namespace.

    Returns
    -------
    logging.Logger
        The initialized root logger.
    """
    logger = logging.getLogger("")
    logger.setLevel(logging.DEBUG)

    if parsed_arguments.verbose is None:
        logger_level = logging.WARNING
    elif parsed_arguments.verbose == 1:
        logger_level = logging.INFO
    else:
        logger_level = logging.DEBUG

    if parsed_arguments.logfile is not None:
        handler = logging.FileHandler(parsed_arguments.logfile)
    else:
        handler = logging.StreamHandler()
    logger.setLevel(logger_level)
    handler.setLevel(logger_level)
    formatter = logging.Formatter("%(asctime)s - %(levelname)s - %(name)s: %(message)s")
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    # Don't collect information about where calls were made from (slows down PyPy)
    logging._srcfile = None
    return logger


def read_config(config_file: str) -> ConfigParser:
    """
    Parse the configuration file.

    Parameters
    ----------
    config_file : str
        The filename of the configuration file.

    Returns
    -------
    configparser.ConfigParser
        The parsed configuration file.

    Raises
    ------
    RuntimeError
        If the configuration file does not exist.
    """
    config = ConfigParser()
    if not config.read(config_file):
        raise RuntimeError("Given configuration file does not exist.")
    return config


def print_start_message() -> None:
    """Print the start message which includes the copyright."""
    print("JeLLFysh (version {0}) - a Python application for all-atom event-chain Monte Carlo - "
          "https://github.com/jellyfysh".format(version.__version__))
    print("Copyright (C) 2019, 2022 The JeLLyFysh organization")


def main() -> None:
    """
    Use the command line arguments to run the JeLLyFysh application.

    First the command line arguments are parsed, and then the logging is set up. Afterwards, the configuration file
    specified in the command line is parsed. Based on the configuration file, the setting package is initialized and the
    mediator is constructed by the JeLLyFysh factory.
    The run method of the mediator is executed until an EndOfRun exception is raised. This invokes the post_run method
    of the mediator and ends the run of the application.
    """
    print_start_message()

    args = parse_options(sys.argv[1:])
    logger = set_up_logging(args)

    logger.info("Run identification hash: {0}".format(get_uuid()))
    logger.info("Underlying platform (determined via platform.platform(aliased=True): {0}"
                .format(platform.platform(aliased=True)))

    logger.info("Setting up the run based on the configuration file {0}.".format(args.config_file))
    config = read_config(args.config_file)
    factory.build_from_config(config, to_camel_case(config.get("Run", "setting")), "jellyfysh.setting")
    mediator = factory.build_from_config(config, to_camel_case(config.get("Run", "mediator")),
                                         "jellyfysh.mediator")
    used_sections = factory.used_sections
    for section in config.sections():
        if section not in used_sections and section != "Run":
            logger.warning("The section {0} in the .ini file has not been used!".format(section))

    logger.info("Running the event-chain Monte Carlo simulation.")
    start_time = time.time()
    try:
        mediator.run()
    except EndOfRun:
        logger.info("EndOfRun exception has been raised.")
    end_time = time.time()

    logger.info("Running the post_run method.")
    mediator.post_run()
    logger.info("Runtime of the simulation: --- %s seconds ---" % (end_time - start_time))


if __name__ == '__main__':
    main()
